rpc/types/block.go

```diff
@@ -159,36 +188,83 @@ func (bnh *BlockNumberOrHash) decodeFromString(input string) error {
 	case BlockParamEarliest:
 		bn := EthEarliestBlockNumber
 		bnh.BlockNumber = &bn
+		return nil
 	case BlockParamLatest:
 		bn := EthLatestBlockNumber
 		bnh.BlockNumber = &bn
+		return nil
 	case BlockParamPending:
 		bn := EthPendingBlockNumber
 		bnh.BlockNumber = &bn
+		return nil
+	case BlockParamFinalized:
+		bn := EthFinalizedBlockNumber
+		bnh.BlockNumber = &bn
+		return nil
+	case BlockParamSafe:
+		bn := EthSafeBlockNumber
+		bnh.BlockNumber = &bn
+		return nil
 	default:
-		// check if the input is a block hash
 		if len(input) == 66 {
 			hash := common.Hash{}
 			err := hash.UnmarshalText([]byte(input))
 			if err != nil {
 				return err
 			}
 			bnh.BlockHash = &hash
-			break
-		}
-		// otherwise take the hex string has int64 value
-		blockNumber, err := hexutil.DecodeUint64(input)
-		if err != nil {
-			return err
+			return nil
 		}
 
-		bnInt, err := ethermint.SafeInt64(blockNumber)
+		blckNum, err := hexutil.DecodeUint64(input)
 		if err != nil {
 			return err
 		}
-
-		bn := BlockNumber(bnInt)
+		if blckNum > math.MaxInt64 {
+			return fmt.Errorf("blocknumber too high")
+		}
+		bn := BlockNumber(blckNum)
 		bnh.BlockNumber = &bn
+		return nil
+	}
+}
+
+func (bnh *BlockNumberOrHash) Number() (BlockNumber, bool) {
+	if bnh.BlockNumber != nil {
+		return *bnh.BlockNumber, true
+	}
+	return BlockNumber(0), false
+}
+
+func (bnh *BlockNumberOrHash) String() string {
+	if bnh.BlockNumber != nil {
+		return strconv.Itoa(int(*bnh.BlockNumber))
```

Comment:
I know it might be redundant as it is check in the decoding but it might be better to play it safe here and check that blockNumber does not overflow the integer type